src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -235,7 +235,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend_reply(MPIR_Context_id_t context_i
         dt = ucp_dt_make_iov();
         total_sz = 2;
     } else {
-        send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
+        MPL_gpu_malloc_host((void **) &send_buf, data_sz + am_hdr_sz + sizeof(ucx_hdr));
```

Comment:
Note: for non-contig datatypes in UCX we can actually allocate a device buffer instead of device registered host buffer and do packing from non-contig device buffer to contig device buffer. For small buffers we still do packing into device registered host buffer.

Reply:
Do we need this fallback path for UCX at all?  We use the pack callback function to do packing for us, when UCX cannot move data automatically.  The pack function can then decide to pack it on the device buffer.  I don't see any case where we need to pack to the host buffer directly.

Reply:
Here are the pack/unpack callback functions that I'm referring to: https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/netmod/ucx/ucx_datatype.c

Reply:
you are right I missed those.

Reply:
However, in this case we still need to do packing because we have the ucx and am headers as well to be sent. We can avoid packing ucx and am headers with payload if payload is in contig buffer by using iov. My idea here was to convert the non-contig device buffer into contig device buffer and pass this with the iov. It also looks like the MPICH pack/unpack callbacks are used for generic types only, which would work for the direct netmod path but not for the AM path.

Reply:
Why would the generic type not work for AM?

Reply:
In AM we either use contig or iov because we have headers too

Reply:
After doing some experiment with AM fallbacks on GPU cluster I have noticed UCX does internal packing of IOV instead of managing segments separately. This means the iov handling does a memcpy of the segments and causes failures if one of them is in GPU memory. I had done some tests about a year ago using UCX + am-only and I don't remember this happening.